File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/MockitoOngoingReturningClause.java
package fr.xgouchet.mesmaeker.mockito;

import org.mockito.Mockito;

import java.util.function.Function;

/**
 * @author Xavier F. Gouchet
 */
public class MockitoOngoingReturningClause<T, O> implements OngoingReturningClause<T, O> {

    private final T subject;
    private final O firstResult;
    private final O[] nextResults;

    @SuppressWarnings("unchecked")
    public MockitoOngoingReturningClause(T subject, O toBeReturned, O... toBeReturnedNext) {
        this.subject = subject;
        firstResult = toBeReturned;
        nextResults = toBeReturnedNext;
    }

    @Override
    public OngoingReturningClause<T, O> when(Function<T, O> function) {
        T stubber;
        if ((nextResults == null) || (nextResults.length == 0)) {
            stubber = Mockito.doReturn(firstResult).when(subject);
        } else {
            stubber = Mockito.doReturn(firstResult, (Object[]) nextResults).when(subject);
        }
        function.apply(stubber);
        return this;
    }
}

File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/MockitoOngoingReturningClause.kt
package fr.xgouchet.mesmaeker.mockito

import org.mockito.Mockito

import java.util.function.Function

/**
 * @author Xavier F. Gouchet
 */
class MockitoOngoingReturningClause<T, O>(private val subject: T,
                                          private val value: O,
                                          vararg values: O)
    : OngoingReturningClause<T, O> {

    private val nextValues: List<O> = listOf(*values)

    override fun `when`(function: Function<T, O>): OngoingReturningClause<T, O> {
        return whenever(function)
    }

    override fun whenever(function: Function<T, O>): OngoingReturningClause<T, O> {
        val stubber: T

        if (nextValues.isEmpty()) {
            stubber = Mockito.doReturn(value).`when`(subject)
        } else {
            // ugly trick to expand as varargs
            val array = nextValues.map { it as Any }.toTypedArray()
            stubber = Mockito.doReturn(value, *array).`when`(subject)
        }
        function.apply(stubber)
        return this
    }


}
